src/keyRotation/keyRotator.ts

```diff
@@ -37,43 +37,70 @@ export class KeyRotator {
      * @param user the IAM User to get the Access Keys for
      */
     private getExistingKeys = (user: string): Promise<AccessKeyMetadata[]> => {
-        console.log(`Retrieving existing keys for User ${user}`);
+        console.log(`Retrieving existing keys for ${user}`);
         const params: ListAccessKeysRequest = {
             UserName: user,
         };
 
         return this.iam.listAccessKeys(params)
             .promise()
             .then((data) => {
-                console.log(`Retrieved the following keys for User ${user}: ${JSON.stringify(data.AccessKeyMetadata)}`);
+                console.log(`Retrieved the following keys for ${user}: ${JSON.stringify(data.AccessKeyMetadata)}`);
                 return Promise.resolve(data.AccessKeyMetadata);
             });
     }
 
     /**
-     * Rotate the given Access Keys for the given IAM User.
+     * Performs the core key rotation steps and attempts to self-heal if there are any errors.
      * @param user the IAM User that the Access Keys belong to
      * @param keys the Access Keys to rotate
      */
     private performKeyRotation = (user: string, keys: AccessKeyMetadata[]) => {
+        return this.performCoreKeyRotation(user, keys)
+            .catch((err) => {
+                console.error(`There was an error during key rotation: ${JSON.stringify(err)}`);
+                return this.selfHeal(user, keys);
+            });
+    }
+
+    /**
+     * Performs the core key rotation steps: creating a new key, propagating it as required and
+     * deleting any old keys.
+     * @param user the IAM User that the Access Keys belong to
+     * @param keys the Access Keys to rotate
+     */
+    private performCoreKeyRotation = (user: string, keys: AccessKeyMetadata[]) => {
+        console.log(`Beginning key rotation.`);
         return this.createNewKey(user)
             .then((key) => this.handleNewKey(user, key))
-            .then(() => this.deleteKeys(user, keys))
-            .catch((err) => {
-                // Try to self-heal by removing any inactive keys but still throw an error
-                // as we haven't created/handled the new key correctly
-                console.log(`Attempting to delete inactive keys`);
-                return this.deleteKeys(user, keys, (key) => key.Status === INACTIVE)
-                    .then(() => Promise.reject(err));
+            .then(() => {
+                console.log(`Deleting old keys.`);
+                return this.deleteKeys(user, keys);
+            })
+            .then(() => {
+                console.log(`Key rotation for ${user} completed succesfully.`);
+                return Promise.resolve();
```

Comment:
I don't think this line is needed

Reply:
Yep, I've removed this return and gone through the rest of the code and removed any explicit calls to `Promise.resolve()` - replaced with `return <value>` - and `Promise.reject(err)` - replaced with `throw err`.